### PR TITLE
Rr/rali dataloader_mt utility

### PR DESCRIPTION
**Pull request by rrawther**
Added rali_dataloader_mt and timing info for shuffle

### CODE REVIEW

File: utilities/rali/rali_dataloader_mt/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+This application demonstrates a basic usage of RALI's C API to use sharded data_loader  in a multithreaded application.
+<p align="center"><img width="90%" src="../../docs/images/image_augmentation.png" /></p>
+
+## Build Instructions
+
+### Pre-requisites
+* Ubuntu Linux, [version `16.04` or later](https://www.microsoft.com/software-download/windows10)
+* RALI library (Part of the MIVisionX toolkit)
+* [OpenCV 3.4+](https://github.com/opencv/opencv/releases/tag/3.4.0)
+* Radeon Performance Primitives (RPP)
```
Comment (kiritigowda) [inline utilities/rali/rali_dataloader_mt/README.md:11]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979)

**Comment by kiritigowda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Clones added
============
- utilities/rali/rali_dataloader_mt/rali_dataloader_mt.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979&bid=18801898)

File: utilities/rali/rali_dataloader_mt/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+### build
+  ````
```
Comment (kiritigowda) [inline utilities/rali/rali_dataloader_mt/README.md:14]: Codacy found an issue: [[fenced-code-flag] Missing code language flag](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979)

File: utilities/rali/rali_dataloader_mt/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# rali_dataloader_mt application
+This application demonstrates a basic usage of RALI's C API to use sharded data_loader  in a multithreaded application.
+<p align="center"><img width="90%" src="../../docs/images/image_augmentation.png" /></p>
+
+## Build Instructions
+
+### Pre-requisites
+* Ubuntu Linux, [version `16.04` or later](https://www.microsoft.com/software-download/windows10)
+* RALI library (Part of the MIVisionX toolkit)
```
Comment (kiritigowda) [inline utilities/rali/rali_dataloader_mt/README.md:9]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979)

File: utilities/rali/rali_dataloader_mt/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
+# rali_dataloader_mt application
+This application demonstrates a basic usage of RALI's C API to use sharded data_loader  in a multithreaded application.
+<p align="center"><img width="90%" src="../../docs/images/image_augmentation.png" /></p>
+
+## Build Instructions
+
+### Pre-requisites
+* Ubuntu Linux, [version `16.04` or later](https://www.microsoft.com/software-download/windows10)
```
Comment (kiritigowda) [inline utilities/rali/rali_dataloader_mt/README.md:8]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979)

File: utilities/rali/rali_dataloader_mt/README.md [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+* RALI library (Part of the MIVisionX toolkit)
+* [OpenCV 3.4+](https://github.com/opencv/opencv/releases/tag/3.4.0)
```
Comment (kiritigowda) [inline utilities/rali/rali_dataloader_mt/README.md:10]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979)

File: utilities/rali/rali_dataloader_mt/README.md (hunk 1 above)
Comment (rrawther) [reply utilities/rali/rali_dataloader_mt/README.md]: Pushed the changes to README. I found that all other README in rali utilities will have the same issue

**Comment by kiritigowda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Clones added
============
- utilities/rali/rali_dataloader_mt/rali_dataloader_mt.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979&bid=18801898)

File: utilities/rali/rali_dataloader_mt/README.md [hunk 6]
```diff
@@ -0,0 +1,25 @@
+# rali_dataloader_mt application
+This application demonstrates a basic usage of RALI's C API to use sharded data_loader  in a multithreaded application.
+<p align="center"><img width="90%" src="../../docs/images/image_augmentation.png" /></p>
+
+## Build Instructions
+
+### Pre-requisites
+*   Ubuntu Linux, [version `16.04` or later] (https://ubuntu.com/download)
```
Comment (kiritigowda) [inline utilities/rali/rali_dataloader_mt/README.md:8]: Codacy found an issue: [[no-shortcut-reference-link] Use the trailing `[]` on reference links](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979)

File: utilities/rali/rali_dataloader_mt/README.md [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+*   RALI library (Part of the MIVisionX toolkit)
+*   [OpenCV 3.4+](https://github.com/opencv/opencv/releases/tag/3.4.0)
+*   Radeon Performance Primitives (RPP)
+
+### build
+  ````
```
Comment (kiritigowda) [inline utilities/rali/rali_dataloader_mt/README.md:14]: Codacy found an issue: [[fenced-code-flag] Missing code language flag](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979)

File: utilities/rali/rali_dataloader_mt/README.md (hunk 6 above)
Comment (kiritigowda) [inline utilities/rali/rali_dataloader_mt/README.md:8]: Codacy found an issue: [[no-literal-urls] Don’t use literal URLs without angle brackets](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979)

Comment (kiritigowda) [inline utilities/rali/rali_dataloader_mt/README.md:8]: Codacy found an issue: [[no-undefined-references] Found reference to undefined definition](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979)

Comment (rrawther) [reply utilities/rali/rali_dataloader_mt/README.md]: what is it complaining about? the reference link is valid

**Comment by kiritigowda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Clones added
============
- utilities/rali/rali_dataloader_mt/rali_dataloader_mt.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979&bid=18801898)

File: utilities/rali/rali_dataloader_mt/README.md [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+## Build Instructions
+
+### Pre-requisites
+*   Ubuntu Linux 16.04/18.04
+*   RALI library (Part of the MIVisionX toolkit)
+*   [OpenCV 3.4+](https://github.com/opencv/opencv/releases/tag/3.4.0)
+*   Radeon Performance Primitives (RPP)
+
+### build
+  ````
```
Comment (kiritigowda) [inline utilities/rali/rali_dataloader_mt/README.md:14]: Codacy found an issue: [[fenced-code-flag] Missing code language flag](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979)

**Comment by kiritigowda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 5
           

Clones added
============
- utilities/rali/rali_dataloader_mt/rali_dataloader_mt.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979&bid=18801898)

File: rali/rali_pybind/amd/rali/plugin/pytorch.py [hunk 9]
```diff
@@ -99,6 +99,9 @@ def __iter__(self):
     def __len__(self):
         return self.len
 
+    def __del__(self):
```
Comment (kiritigowda) [inline rali/rali_pybind/amd/rali/plugin/pytorch.py:102]: Codacy found an issue: [Missing docstring in magic method (D105)](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979)

File: utilities/rali/rali_dataloader_mt/README.md (hunk 5 above)
Comment (kiritigowda) [inline utilities/rali/rali_dataloader_mt/README.md:10]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979)

File: utilities/rali/rali_dataloader_mt/README.md (hunk 4 above)
Comment (kiritigowda) [inline utilities/rali/rali_dataloader_mt/README.md:8]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979)

File: utilities/rali/rali_dataloader_mt/README.md (hunk 3 above)
Comment (kiritigowda) [inline utilities/rali/rali_dataloader_mt/README.md:9]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979)

File: utilities/rali/rali_dataloader_mt/README.md (hunk 1 above)
Comment (kiritigowda) [inline utilities/rali/rali_dataloader_mt/README.md:11]: Codacy found an issue: [[list-item-indent] Incorrect list-item indent: add 2 spaces](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979)

**Comment by LakshmiKumar23** [comment]
@rrawther When I give a sub folder of the imagenet datadet as <path-to-images>, I see this:
```
>>> Running on CPU>>> Running on CPU

[INF]  {compute_optimum_internal_batch_size}  Can run 32 threads simultaneously on this machine
[INF]  {compute_optimum_internal_batch_size}  User batch size 16 Internal batch size set to 16
RALI created context for <02 >
[INF]  {raliJpegFileSourceSingleShard}  Internal buffer size width = 1024 height = 1024 depth = 3
[INF]  {compute_optimum_internal_batch_size}  Can run 32 threads simultaneously on this machine
[INF]  {compute_optimum_internal_batch_size}  User batch size 16 Internal batch size set to 16
RALI created context for <12 >
[INF]  {raliJpegFileSourceSingleShard}  Internal buffer size width = 1024 height = 1024 depth = 3
Remaining images 656
Augmented copies count 1
output width 224 output height 3584 color planes 3 n 1
[INF]  {output_routine}  Output routine started with 656 to load
Remaining images 656
Augmented copies count 1
output width 224 output height 3584 color planes 3 n 1
[INF]  {output_routine}  Output routine started with 656 to load
terminate called after throwing an instance of 'RaliException'
  what():   { meta_data } No meta data has been loaded
Aborted (core dumped)
```

**Comment by kiritigowda** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Clones added
============
- utilities/rali/rali_dataloader_mt/rali_dataloader_mt.cpp  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=5745979&bid=18801898)

**Comment by LakshmiKumar23** [comment]
@rrawther I still get this. ANy other steps to follow before running the app?
```
./rali_dataloader_mt ~/work/mlperf-training/imagenet_10dir/train/n01978287 0 1 224 224 16 0 0
>>> Running on CPU
[INF]  {compute_optimum_internal_batch_size}  Can run 32 threads simultaneously on this machine
[INF]  {compute_optimum_internal_batch_size}  User batch size 16 Internal batch size set to 16
RALI created context for <01 >
[INF]  {raliJpegFileSourceSingleShard}  Internal buffer size width = 224 height = 224 depth = 3
Remaining images 1312
[INF]  {output_routine} Augmented copies count 1
 output width 224 output height 3584 color planes 3 n 1
Output routine started with 1312 to load
terminate called after throwing an instance of 'RaliException'
  what():   { meta_data } No meta data has been loaded
Aborted (core dumped)
```

**Comment by rrawther** [comment]
@LakshmiKumar23 : I didn't get that error. But I have seen it with image_augmentation app as well. I think it is timing related. Let me see if I can disable that code

**Comment by rrawther** [comment]
@LakshmiKumar23 : Can you please try again by giving one more parameter at the end for display (0/1)

**Comment by rrawther** [comment]
@kiritigowda and @LakshmiKumar23 can we me merge this PR

**Comment by LakshmiKumar23** [comment]
@rrawther @kiritigowda tried with cpu and gpu and different batch sizes. Runs successfully. I couldn't see the display though. Maybe the dataset was too small and it closed before I could view it.

**Comment by rrawther** [comment]
@LakshmiKumar23 : I have disabled the display since it was not working from multithreaded app. It is just for testing the multithreading with sharding. Display is not needed. Please merge it if it ran

**Comment by LakshmiKumar23** [comment]
@kiritigowda Ready to be merged
